### PR TITLE
feat(tools-react-native): add function to get available platforms

### LINKED ISSUE #925: Figure out a way to dynamically get available platforms
### What happened?

We are currently hard-coding the list of platform-to-npm-package mapping because `react-native config` is too slow for our use cases.

https://github.com/microsoft/rnx-kit/blob/1a2cf6720d562ad97d5fdb02a9c5bab83192ba01/packages/tools-react-native/src/platform.ts#L6-L13

We should figure out alternative ways to populate this list.

### Affected Package

@rnx-kit/tools-react-native

### Version

main

### Which platforms are you seeing this issue on?

- [ ] Android
- [ ] iOS
- [ ] macOS
- [ ] Windows

### System Information

```true
n/a
```

### Steps to Reproduce

n/a

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### PR DESCRIPTION
**Pull request by tido64**
### Description

Added a function to get available React Native platforms.

Resolves #925

### Test plan

Tests were added. CI should pass.

### CODE REVIEW

File: packages/metro-resolver-symlinks/src/resolver.ts [hunk 1]
```diff
@@ -33,7 +43,7 @@ export const remapReactNativeModule: ModuleResolver = (
   moduleName,
   platform
 ) => {
-  const platformImpl = AVAILABLE_PLATFORMS[platform];
+  const platformImpl = getPlatformImplementations()[platform];
```
Comment (tido64) [inline packages/metro-resolver-symlinks/src/resolver.ts:46]: @afoxman: Can we also replace the occurrence in `typescript-react-native-resolver`?

https://github.com/microsoft/rnx-kit/blob/2e2f8b94ad6ede2964f282a65e0c4e622007e3e4/packages/typescript-react-native-resolver/src/react-native-package-name.ts#L21

File: packages/tools-react-native/test/__fixtures__/available-platforms/package.json [hunk 2]
```diff
@@ -0,0 +1,12 @@
+{
+  "name": "@rnx-kit/tools-react-native/available-platform-tests",
+  "version": "0.0.0-dev",
+  "dependencies": {
+    "react-native": "0.0.0-dev",
+    "react-native-codegen": "0.0.0-dev",
+    "react-native-macos": "0.0.0-dev",
+    "react-native-tvos": "0.0.0-dev",
+    "react-native-win32": "0.0.0-dev",
```
Comment (afoxman) [inline packages/tools-react-native/test/__fixtures__/available-platforms/package.json:9]:
```suggestion
    "@office-iss/react-native-win32": "0.0.0-dev",
```

This package is published under the `@office-iss` scope.

File: packages/metro-resolver-symlinks/test/__fixtures__/remap-platforms/package.json [hunk 3]
```diff
@@ -0,0 +1,10 @@
+{
+  "name": "@rnx-kit/metro-resolver-symlinks/remap-tests",
+  "version": "0.0.0-dev",
+  "dependencies": {
+    "react-native": "0.0.0-dev",
+    "react-native-macos": "0.0.0-dev",
+    "react-native-win32": "0.0.0-dev",
```
Comment (afoxman) [inline packages/metro-resolver-symlinks/test/__fixtures__/remap-platforms/package.json:7]:
```suggestion
    "@office-iss/react-native-win32": "0.0.0-dev",
```

This package is published under the `@office-iss` scope.

File: packages/metro-resolver-symlinks/src/resolver.ts [hunk 4]
```diff
@@ -1,8 +1,18 @@
 import { isFileModuleRef, parseModuleRef } from "@rnx-kit/tools-node";
-import { AVAILABLE_PLATFORMS } from "@rnx-kit/tools-react-native";
+import { getAvailablePlatforms } from "@rnx-kit/tools-react-native";
 import * as path from "path";
 import type { MetroResolver, ModuleResolver } from "./types";
 
+const getPlatformImplementations = (() => {
```
Comment (afoxman) [inline packages/metro-resolver-symlinks/src/resolver.ts:6]: Why not make this cached function also available as part of the tools library?

File: packages/metro-resolver-symlinks/test/__fixtures__/remap-platforms/node_modules/react-native-win32/package.json [hunk 5]
```diff
@@ -0,0 +1,4 @@
+{
+  "name": "react-native-win32",
```
Comment (afoxman) [inline packages/metro-resolver-symlinks/test/__fixtures__/remap-platforms/node_modules/react-native-win32/package.json:2]: This package should include the `@office-iss` scope in its name and path, as that is how it is published.

File: packages/metro-resolver-symlinks/test/__fixtures__/remap-platforms/node_modules/react-native-win32/react-native.config.js [hunk 6]
```diff
@@ -0,0 +1,10 @@
+module.exports = {
```
Comment (afoxman) [inline packages/metro-resolver-symlinks/test/__fixtures__/remap-platforms/node_modules/react-native-win32/react-native.config.js:1]: This package should include the `@office-iss` scope in its path, as that is how it is published.

File: packages/tools-react-native/src/platform.ts [hunk 7]
```diff
@@ -34,6 +37,54 @@ export function expandPlatformExtensions(
   return expanded;
 }
 
+/**
+ * Returns a map of available React Native platforms.
```
Comment (afoxman) [inline packages/tools-react-native/src/platform.ts:41]: Can you add @jsdoc tags here to fully document the API?

File: packages/tools-react-native/test/__fixtures__/available-platforms/node_modules/react-native-win32/package.json [hunk 8]
```diff
@@ -0,0 +1,4 @@
+{
```
Comment (afoxman) [inline packages/tools-react-native/test/__fixtures__/available-platforms/node_modules/react-native-win32/package.json:1]: This package should include the `@office-iss` scope in its name and path, as that is how it is published.

File: packages/tools-react-native/test/__fixtures__/available-platforms/node_modules/react-native-win32/react-native.config.js [hunk 9]
```diff
@@ -0,0 +1,10 @@
+module.exports = {
```
Comment (afoxman) [inline packages/tools-react-native/test/__fixtures__/available-platforms/node_modules/react-native-win32/react-native.config.js:1]: This package should include the `@office-iss` scope in its path, as that is how it is published.

File: packages/metro-resolver-symlinks/src/resolver.ts (hunk 1 above)
Comment (afoxman) [reply packages/metro-resolver-symlinks/src/resolver.ts:46]: Yes, with some caveats.

The whole thing is based on process.cwd() as the package root. To make both CLI and programmatic use work, you should pass in an explicit root from the calling code:

```
host.getCurrentDirectory()
```

Also, please use a cached version so the search only runs once per process.

File: packages/tools-react-native/test/platform.test.ts [hunk 10]
```diff
@@ -1,5 +1,7 @@
+import * as path from "path";
 import {
   expandPlatformExtensions,
+  getAvailablePlatformsUncached as getAvailablePlatforms,
```
Comment (afoxman) [inline packages/tools-react-native/test/platform.test.ts:4]: Changing the name like this makes the test bodies misleading, since both are valid API names.

**Review by afoxman — APPROVED** [review]

**Pull request merged**